libraries/esp8266/examples/LowPowerDemo/LowPowerDemo.ino

```diff
@@ -0,0 +1,338 @@
+  Serial.begin(115200);
+  Serial.print(F("\nReset reason = "));
+  String resetCause = ESP.getResetReason();
+  Serial.println(resetCause);
+  if ((resetCause == "External System") || (resetCause == "Power on")) {
+    Serial.println(F("I'm awake and starting the Low Power tests"));
+    resetCount = 5;
+    updateRTC();  // if external reset, wipe the RTC memory and start all over
+  }
+
+  // Read struct from RTC memory
+  if (ESP.rtcUserMemoryRead(64, (uint32_t*) &rtcData, sizeof(rtcData))) {
+    uint32_t crcOfData = crc32((uint8_t*) &rtcData.data[0], sizeof(rtcData.data));
+    if (crcOfData != rtcData.crc32) {  // if the CRC is invalid
+      resetCount = 0;  // set first test run since power on or external reset
+    } else {
+      resetCount = rtcData.data[3];  // read the previous reset count
+    }
+  }
+}  // end of setup()
+
+void loop() {
+  if (resetCount == 0) {  // first reset, else skip down to the next Deep Sleep test
+    // 1st test - running with WiFi unconfigured, reads ~67 mA minimum
+    Serial.println(F("\n1st test - running with WiFi unconfigured"));
+    float volts = ESP.getVcc();
+    Serial.printf("The internal VCC reads %1.3f volts\n", volts / 1000);
+    Serial.println(F("press the button to continue"));
+    waitPushbutton(false, blinkDelay);
+
+    // 2nd test - Automatic Modem Sleep 7 seconds after WiFi is connected (LED flashes)
+    Serial.println(F("\n2nd test - Automatic Modem Sleep"));
+    Serial.println(F("connecting WiFi, please wait until the LED blinks"));
+    init_WiFi();
+    if (WiFi.localIP()) {  // won't go into Automatic Sleep without an active WiFi connection
+      Serial.println(F("The amperage will drop in 7 seconds."));
+      volts = ESP.getVcc();
+      Serial.printf("The internal VCC reads %1.3f volts\n", volts / 1000);
+      Serial.println(F("press the button to continue"));
+      waitPushbutton(true, longDelay);
+    } else {
+      Serial.println(F("no WiFi connection, test skipped"));
+    }
+
+    // 3rd test - Forced Modem Sleep
+    Serial.println(F("\n3rd test - Forced Modem Sleep"));
+    WiFi.forceSleepBegin();
+    delay(10);  // it doesn't always go to sleep unless you delay(10); yield() wasn't reliable
+    // WiFi.mode(WIFI_SHUTDOWN);  // alternate method of Forced Modem Sleep
+    volts = ESP.getVcc();
+    Serial.printf("The internal VCC reads %1.3f volts\n", volts / 1000);
+    Serial.println(F("press the button to continue"));
+    waitPushbutton(true, longDelay);
+
+    // 4th test - Automatic Light Sleep
+    Serial.println(F("\n4th test - Automatic Light Sleep"));
+    Serial.println(F("reconnecting WiFi"));
+    Serial.println(F("it will be in Automatic Light Sleep once WiFi connects (LED blinks)"));
+    digitalWrite(LED, LOW);  // visual cue that we're reconnecting
+    WiFi.setSleepMode(WIFI_LIGHT_SLEEP, 5);  // Automatic Light Sleep
+    WiFi.forceSleepWake();  // reconnect with previous STA mode and connection settings
+    uint32_t wifiStart = millis();
```

Comment:
Please use polledTimeout here to not reinvent the wheel, like so:
```cpp
using oneShot=esp8266::polledTimeout::oneShot; //shorthand for the oneShot type (this is the same as a typedef)
oneShot timeout(wifiTimeout);
while(!WiFi.localIP && !timeout) {
  delay(50);
}
```

Reply:
Even simpler: I replaced various delay()s in the WiFi inits with yield().